Fix #9591 by filling default labels on boolean fields from the API

dhilst opened the pull request:
Whenc creating a boolean field the UI provides default values for Yes/No labels
but the API doesn't and no error were given until user navigate to a page that load lead fields.

This commit mimics the UI behavior on the API by filling default Yes and No labels automatically

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" for all features, enhancements and bug fixes (until 3.3.0 is released) <!-- see below -->
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | no
| Related user documentation PR URL      | None... <!-- required for new features -->
| Related developer documentation PR URL | https://developer.mautic.org/#create-field
| Issue(s) addressed                     | Fixes #9591
<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:

When creating a boolean custom field in the UI, the Yes/No labels are rendered with "Yes" and "No" default values and sent to backend. But on when creating the same from the API, if no `..., properties: { "yes": "Yes", "no: "No" } ... }` is provided the field *is* created, by with blank "yes" and "no" properties. This then crashes some pages when ListBundle is loaded with classic Undefined index error.

<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

* Enable API on Settings
* Run this curl (edit url, user and password)
  ```
  curl -v -L \
    -H 'Content-Type: application/json' \
    -u admin:xxxxxxxxx \
    http://127.0.0.1:8081/index_dev.php/api/fields/contact/new \
    -d '{ "label": "foo", "alias": "foo", "type": "boolean" }'
  ```
* Navigate to Segments and try to create a new segment, you should get an alert, 
  <img width="294" alt="image" src="https://user-images.githubusercontent.com/392820/113522454-e96f5800-9576-11eb-888b-b73913b8e40b.png">

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

dhilst commented:
I didn't get why CS Fixer is failing :(, can I run it in the command file to check? Also I saw that this decreases the coverage, I'm reading other tests to figure out how to test this.

dhilst commented:
I added some tests, I would like to test this again by hand because I moved the fix code to it's own function, but now features is in php8 and I will have to install it :)

dhilst commented:
I was looking on this. It seems that the commits are wrong authored I need to amend it. But I don't understand the fixer problem and the coverage problem. I added tests but coverage still reports -0.01of coverage is there anyway to validate this on my machine prior pusshing!?

escopecz commented on app/bundles/ApiBundle/Controller/CommonApiController.php:
```diff
@@ -861,6 +862,22 @@ protected function getEntityDefaultProperties($entity)
+        // Fill boolean labels if they are missing, otherwise they lead to
+        // undefined index error on ListModel
+        // See https://github.com/mautic/mautic/issues/9591
+        echo "outside if\n";
```
This is probably a forgotten debug line?

escopecz commented on app/bundles/ApiBundle/Controller/CommonApiController.php:
```diff
@@ -861,6 +862,22 @@ protected function getEntityDefaultProperties($entity)
         return $properties;
     }
 
+    protected function sanitizeProperties(object $entity, array $parameters)
```
I think this shouldn't be located directly in the CommonApiController as all entities use it. But only 2 (contact, company) have custom fields. How about to implement it into https://github.com/mautic/mautic/blob/features/app/bundles/LeadBundle/Controller/Api/CustomFieldsApiControllerTrait.php ?

escopecz commented on app/bundles/ApiBundle/Controller/CommonApiController.php:
```diff
@@ -861,6 +862,22 @@ protected function getEntityDefaultProperties($entity)
+        echo "outside if\n";
+        if ($entity instanceof LeadField &&
+            (isset($parameters['type']) && 'boolean' === $parameters['type'])) {
+            $parameters['properties']        = isset($parameters['properties']) ? $parameters['properties'] : [];
```
Can be shorter
```suggestion
            $parameters['properties']        = $parameters['properties'] ?? [];
```

escopecz commented on app/bundles/ApiBundle/Tests/Controller/CommonApiControllerTest.php:
```diff
@@ -86,9 +88,68 @@ public function testgetWhereFromRequestWithSomeWhere()
     }
 
-    protected function getResultFromProtectedMethod($method, array $args, Request $request = null)
+    public function testgetEntityDefaultProperties()
```
Consider modifying https://github.com/mautic/mautic/blob/features/app/bundles/LeadBundle/Tests/Controller/Api/LeadApiControllerFunctionalTest.php#L20 instead

escopecz reviewed (changes requested):
Good catch! Please take a look at the suggestions.

escopecz commented:
Don't worry about the CS Fixer. We'll fix that.

dhilst commented:
Hi @escopecz thanks for the feedback I will update the PR  if your suggestions (y)

RCheesley commented:
Hi @dhilst did you get a chance to take the feedback into consideration? I noticed that we now have a conflict to address, can you take a look please?

dhilst commented:
Hi @RCheesley yeah I had unpushed commits on this. I have committed the code but I need to setup php8 on my machine to test it and got busy in this middle time. I can solve the conflicts, don't worry I will create a time for this in the next week.

Regards

dhilst committed:
Fix #9591 by filling default labels on boolean fields from the API

The UI does this automatically by renderizing the fields in the frontend
but the API doesn't and no error were given until user navigate to a
page that load lead fields.

This commit mimics the UI behavior on the API by filling default Yes
and No labels automatically

dhilst commented:
Hi @escopecz I update the PR with your recomendations, I put that `sanitizeProperties` on `FieldsApiController` since the trait wasn't being used there and no other class is using this method. Also I added the tests to `FieldApiControllerFunctionalTest`. Can you review please? 

Regards,

RCheesley committed:
Merge branch 'features' into fix/field-rest-accept-missing-labels

kuzmany reviewed (approved):
Works 👍 
Code looks good.
Unit tests passed
